connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java

```diff
@@ -0,0 +1,155 @@
+                            t
+                    );
+                }
+                return null;
+            }
+        );
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    protected <T> PluginDesc<T> pluginDesc(Class<? extends T> plugin, String version, ClassLoader loader) {
+        return new PluginDesc(plugin, version, loader);
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <T> SortedSet<PluginDesc<T>> getServiceLoaderPluginDesc(Class<T> klass, ClassLoader loader) {
+        SortedSet<PluginDesc<T>> result = new TreeSet<>();
+        ServiceLoader<T> serviceLoader = ServiceLoader.load(klass, loader);
+        for (Iterator<T> iterator = serviceLoader.iterator(); iterator.hasNext(); ) {
+            try (LoaderSwap loaderSwap = withClassLoader(loader)) {
+                T pluginImpl;
+                try {
+                    pluginImpl = iterator.next();
+                } catch (ServiceConfigurationError t) {
+                    log.error("Failed to discover {}{}", klass.getSimpleName(), reflectiveErrorDescription(t.getCause()), t);
+                    continue;
+                }
+                Class<? extends T> pluginKlass = (Class<? extends T>) pluginImpl.getClass();
+                if (pluginKlass.getClassLoader() != loader) {
+                    log.debug("{} from other classloader {} is visible from {}, excluding to prevent isolated loading",
+                            pluginKlass.getSimpleName(), pluginKlass.getClassLoader(), loader);
+                    continue;
+                }
+                result.add(pluginDesc(pluginKlass, versionFor(pluginImpl), loader));
+            }
+        }
+        return result;
+    }
+
+    protected static <T> String versionFor(T pluginImpl) {
+        try {
+            if (pluginImpl instanceof Versioned) {
+                return ((Versioned) pluginImpl).version();
+            }
+        } catch (Throwable t) {
+            log.error("Failed to get plugin version for " + pluginImpl.getClass(), t);
+        }
+        return DelegatingClassLoader.UNDEFINED_VERSION;
+    }
+
+    protected String reflectiveErrorDescription(Throwable t) {
+        if (t instanceof NoSuchMethodException) {
+            return ": Plugin class must have a no-args constructor, and cannot be a non-static inner class";
+        } else if (t instanceof SecurityException) {
+            return ": Security settings must allow reflective instantiation of plugin classes";
+        } else if (t instanceof IllegalAccessException) {
+            return ": Plugin class default constructor must be public";
+        } else if (t instanceof ExceptionInInitializerError) {
+            return ": Failed to statically initialize plugin class";
+        } else if (t instanceof InvocationTargetException) {
+            return ": Failed to invoke plugin constructor";
+        } else {
+            return "";
+        }
+    }
+
+    public LoaderSwap withClassLoader(ClassLoader loader) {
```

Comment:
Nit: doesn't have to be public
```suggestion
    protected LoaderSwap withClassLoader(ClassLoader loader) {
```